### PR TITLE
Use masked marginal probability

### PR DESCRIPTION
**Pull request by arteemg**

### CODE REVIEW

File: src/protdesign/models/esm2.py [hunk 1]
```diff
@@ -504,6 +510,12 @@ def single_mutation_scan(
                         **mut_scores
                     })
 
+                # Update status callback if provided
+                if status_callback:
+                    progress = (len(mutation_effects) / len(positions)) * 100
+                    status_callback(
```
Comment (thomashopf) [inline src/protdesign/models/esm2.py:517]: Not a valid status callback call - cf definition: 
```
# status, progress (optional), message (optional)
Status = Literal["running", "done", "failed"]
StatusCallback = Callable[[Status, float | None, str | None], Any]
```

File: src/protdesign/models/esm2.py [hunk 2]
```diff
@@ -451,29 +452,34 @@ def single_mutation_scan(
...
-                ref_logits = outputs.logits[0]
+            # For each position to scan
+            for pos in positions:
+                pos_idx = pos - target.first_index
+                wt_aa = instance_seq[pos_idx]
 
-                # Convert logits to log probabilities
-                ref_log_probs = torch.log_softmax(ref_logits, dim=-1)
+                # Create masked sequence - mask the position we want to predict
+                masked_seq = list(instance_seq)
```
Comment (thomashopf) [inline src/protdesign/models/esm2.py]: this variable not used according to linter

File: src/protdesign/models/esm2.py [hunk 3]
```diff
@@ -529,45 +541,79 @@ def score_mutants(
...
+            # Pre-compute masked probabilities for each position that will be mutated
+            position_log_probs = {}
 
-                    for sub in mutant:
-                        pos_idx = sub.pos - target.first_index
-                        wt_aa = instance_seq[pos_idx]
-                        mut_aa = sub.to
+            with torch.no_grad():
+                for pos_idx in mutated_positions:
+                    if status_callback:
```
Comment (thomashopf) [inline src/protdesign/models/esm2.py:565]: not a valid call for status_callback (cf above)

File: src/protdesign/models/esm2.py [hunk 4]
```diff
@@ -529,45 +541,79 @@ def score_mutants(
...
+                    logits = outputs.logits[0, token_pos]
+                    log_probs = torch.log_softmax(logits, dim=-1)
+
+                    # Store the log probabilities for this position
+                    position_log_probs[pos_idx] = log_probs
+
+            # Calculate scores for all mutants using the pre-computed masked probabilities
+            mutant_scores = []
+            for i, mutant in enumerate(mutants):
+                if status_callback:
```
Comment (thomashopf) [inline src/protdesign/models/esm2.py:610]: not a valid call for status_callback (cf above)

File: src/protdesign/models/esm2.py [hunk 5]
```diff
@@ -608,52 +654,82 @@ def score_conditional(
...
-                batch_end = min(
-                    batch_start + self.batch_size, len(seqs)
-                )
-                batch_seqs = seqs[batch_start:batch_end]
-                batch_positions = positions[batch_start:batch_end]
-                batch_indices = list(range(batch_start, batch_end))
-                batch_entities = entities[batch_start:batch_end]
+            # Process each sequence individually for masked-marginals
+            for seq_idx, (seq, pos, entity) in enumerate(zip(seqs, positions, entities)):
+                if status_callback:
```
Comment (thomashopf) [inline src/protdesign/models/esm2.py:690]: not a valid call for status_callback (cf above)

File: src/protdesign/models/esm2.py [hunk 6]
```diff
@@ -608,52 +654,82 @@ def score_conditional(
...
+                                aa_token_id = self.tokenizer.convert_tokens_to_ids(
+                                    aa_tokens[0])
+                                aa_probs[aa] = log_probs[aa_token_id].item()
+
+                    # Store results
+                    conditionals_list.append({
+                        'instance': seq_idx,
+                        'entity': entity,
+                        'pos': pos,
+                        'original_aa': original_aa,
```
Comment (thomashopf) [inline src/protdesign/models/esm2.py]: this column "original_aa" should not be present in dataframe according to specification, it will break things downstream

File: src/protdesign/models/esm2.py [hunk 7]
```diff
@@ -580,7 +626,7 @@ def score_conditional(
     ) -> pd.DataFrame:
```
Comment (thomashopf) [inline src/protdesign/models/esm2.py:651]: realizing that computation for score_conditional is not batched anymore - is there a reason why not? otherwise should definitely have it back in

File: src/protdesign/models/esm2.py [hunk 8]
```diff
@@ -421,7 +422,7 @@ def single_mutation_scan(
         status_callback: StatusCallback | None = None
     ) -> pd.DataFrame:
```
Comment (thomashopf) [inline src/protdesign/models/esm2.py:424]: single mutation scan currently not bached? is there a reason - if not should add for speedup now that we need to compute many marginals

File: src/protdesign/models/esm2.py [hunk 9]
```diff
@@ -529,45 +541,79 @@ def score_mutants(
...
+                    # Mask the position
+                    masked_inputs[0, token_pos] = self.tokenizer.mask_token_id
+
+                    # Forward pass with masked input
+                    masked_model_inputs = {
+                        'input_ids': masked_inputs,
+                        'attention_mask': inputs.attention_mask
+                    }
+
+                    outputs = self.model(**masked_model_inputs)
```
Comment (thomashopf) [inline src/protdesign/models/esm2.py]: also not batching here - cf. other comments - should have this back in if possible
